### PR TITLE
refactoring search

### PR DESCRIPTION
**Pull request by natashaarmbrust**
- Refactors search
- Still need to address scroll view placement when tabbing between Series, Episodes, People
- Stil need to address adding a search delay

### CODE REVIEW

File: Podcast/SearchViewController.swift [hunk 1]
```diff
@@ -9,25 +9,132 @@
...
+        switch self {
+        case .series:
+            return "series"
+        case .episodes:
+            return "episodes"
+        case .people:
+            return "users"
+        }
+    }
+}
```
Comment (dantheli) [inline Podcast/SearchViewController.swift:84]: love this

File: Podcast/TabbedPageViewController.swift [hunk 2]
```diff
@@ -1,308 +1,309 @@
+////
```
Comment (dantheli) [inline Podcast/TabbedPageViewController.swift]: is this file still necessary then

**Review by dantheli — APPROVED** [review]

File: Podcast/SearchTableViewController.swift [hunk 3]
```diff
@@ -1,202 +1,203 @@
...
-    
-    // MARK: SearchITunesHeaderViewDelegate
-    func searchITunesHeaderDidPressSearchITunes(searchITunesHeader: SearchITunesHeaderView) {
-        cellDelegate?.searchTableViewControllerPresentSearchITunes(controller: self)
-    }
-    
-    func searchITunesHeaderDidPressDismiss(searchITunesHeader: SearchITunesHeaderView) {
-        tableView.tableHeaderView = nil
-    }
-}
```
Comment (mindylou) [inline Podcast/SearchTableViewController.swift:202]: you can delete this file

File: Podcast/SearchViewController.swift [hunk 4]
```diff
@@ -9,25 +9,132 @@
...
-        
-        searchResultsController.searchResultsDelegate = self
-        
-        searchController = UISearchController(searchResultsController: searchResultsController)
-        searchController.searchResultsUpdater = searchResultsController
+
+
+        tableViewData = [SearchDataSourceDelegate(searchType: .series), SearchDataSourceDelegate(searchType: .episodes), SearchDataSourceDelegate(searchType: .people)]
+        for data in tableViewData {
+            data.delegate = self
```
Comment (mindylou) [inline Podcast/SearchViewController.swift:110]: nice!

File: Podcast/SearchViewController.swift [hunk 5]
```diff
@@ -43,8 +150,8 @@ class SearchViewController: ViewController, UISearchControllerDelegate, UITableV
         
         navigationItem.titleView = searchController?.searchBar
         
-        //IMPORTANT: Does not implement EmptyStateTableViewDelegate because pastSearch does not have an action button s
-        pastSearchesTableView = EmptyStateTableView(frame: view.frame, type: .pastSearch)
+        //IMPORTANT: Does not implement EmptyStateTableViewDelegate because pastSearch does not have an action button
+        pastSearchesTableView = EmptyStateTableView(frame: CGRect.zero, type: .pastSearch)
```
Comment (mindylou) [inline Podcast/SearchViewController.swift]: I think you can just do `.zero` here

File: Podcast/SearchViewController.swift [hunk 6]
```diff
@@ -72,35 +218,101 @@ class SearchViewController: ViewController, UISearchControllerDelegate, UITableV
...
+    }
+
     //MARK: - Tabbed Search Results Delegate
-    func didTapOnSeriesCell(series: Series) {
+
+    func refreshController() {
+        searchResultsTableView.finishInfiniteScroll()
+        searchResultsTableView.reloadData()
+        searchResultsTableView.stopLoadingAnimation()
+        self.updateTableViewInsetsForAccessoryView()
```
Comment (mindylou) [inline Podcast/SearchViewController.swift]: don't need `self` here

File: Podcast/SearchViewController.swift [hunk 7]
```diff
@@ -155,5 +386,95 @@ class SearchViewController: ViewController, UISearchControllerDelegate, UITableV
...
+    var continueInfiniteScroll: Bool = true
+    var completingNewSearch: Bool = false // two types of searching: inifinite scroll or new search query -> true if a new search query in progress
+
+    weak var delegate: SearchTableViewDelegate?
+
+    init(searchType: SearchType) {
+        self.searchType = searchType
+    }
+
+    func fetchData(query: String, newSearch: Bool = false) {
```
Comment (mindylou) [inline Podcast/SearchViewController.swift:416]: nice!

File: Podcast/SearchViewController.swift [hunk 8]
```diff
@@ -155,5 +386,95 @@ class SearchViewController: ViewController, UISearchControllerDelegate, UITableV
...
+        request.success = { endpoint in
+            guard let results = endpoint.processedResponseValue as? [Any] else { return }
+            self.completingNewSearch = false
+            if results.count < self.pageSize { self.continueInfiniteScroll = false }
+            if newSearch {
+                self.searchResults = results
+            } else {
+                self.searchResults = self.searchResults + results
+            }
+            self.delegate?.refreshController()
```
Comment (mindylou) [inline Podcast/SearchViewController.swift:430]: image
should the refresh controller be consistent with the other loading indicators (the cool one that's used elsewhere)

File: Podcast/SearchViewController.swift [hunk 9]
```diff
@@ -72,35 +218,101 @@ class SearchViewController: ViewController, UISearchControllerDelegate, UITableV
...
+            didTapOnEpisodeCell(episode: episode)
+        } else if let user = object as? User {
+            didTapOnUserCell(user: user)
+        }
+    }
+
+    private func didTapOnSeriesCell(series: Series) {
         addPastSearches()
         let seriesDetailViewController = SeriesDetailViewController(series: series)
         navigationController?.pushViewController(seriesDetailViewController,animated: true)
```
Comment (mindylou) [inline Podcast/SearchViewController.swift:308]: image
going back from a series detail/episode detail to the search causes the search results to disappear for a few seconds/not sure why this is happening

**Review by mindylou — CHANGES_REQUESTED** [review]
just a few small fixes, otherwise looks good!

File: Podcast/TabbedPageViewController.swift (hunk 2 above)
Comment (natashaarmbrust) [reply Podcast/TabbedPageViewController.swift]: nope

File: Podcast/SearchTableViewController.swift (hunk 3 above)
Comment (natashaarmbrust) [reply Podcast/SearchTableViewController.swift:202]: done

File: Podcast/SearchViewController.swift (hunk 5 above)
Comment (natashaarmbrust) [reply Podcast/SearchViewController.swift]: done

File: Podcast/SearchViewController.swift (hunk 6 above)
Comment (natashaarmbrust) [reply Podcast/SearchViewController.swift]: whoops

File: Podcast/SearchViewController.swift (hunk 8 above)
Comment (natashaarmbrust) [reply Podcast/SearchViewController.swift:430]: We've been using this one for refreshing and infinite scroll ... @dantheli and I looked into making it the same indicator and it didn't work / neither of us could figure out why `NVAcitvityIndicatorView` did not show up for an infinite scroll. I the default one looks fine so let's keep it for now.

File: Podcast/SearchViewController.swift (hunk 9 above)
Comment (natashaarmbrust) [reply Podcast/SearchViewController.swift:308]: I can't reproduce this.. any more details?

Comment (natashaarmbrust) [reply Podcast/SearchViewController.swift:308]: nevermind fixed.
